### PR TITLE
[ETHEREUM-CONTRACTS] [GDA] Fixes for GDA 

### LINKED ISSUE #1746: [ETHEREUM-CONTRACTS] [GDA] Audit-related Fixes
## Overview
**What?**
Fixes applied to the GDA-related contracts given the audit by Trail of Bits.

**How?**
A review of the Trail of Bits audit and then application of the fixes if deemed reasonable.
Firstly, an internal review of the fixes applied and internal review of the code again.
Then another three engineer days review of the fixes by Trail of Bits before the final review.

**Why?**
To ensure that any issues found by ToB are resolved to mitigate smart contract security risks.

**Deadline:** 04/12/2023

**UI/UX Design:** N/a

## Acceptance criteria
 - [x] Fix suggestions by Trail of Bits are 
 - [x] Satisfies entire acceptance criteria
 - [ ] Consider communication & visibility
 - [ ] Code review is approved by at least 2 reviewers
 - [ ] Tested by Elvijs/QA & discovered bugs resolved
 - [ ] Documentation updated
 - [ ] Assignee/opened by merges PR
 - [ ] Automated E2E tests for happy path
 - [ ] Code is merged
 - [ ] Observability of critical user path
 - [ ] Steps to access code change as a user

## Links/Resources
n/a

## Readiness criteria
- [x] Descriptive title
- [x] What?
- [x] Why?
- [x] How?
- [x] Acceptance criteria complying with Definition of Done
- [ ] Estimate/Story points
- [x] Deadline: If not a part of an epic
- [x] Priority: Reflected as order on the bord, walking the board from the top and right
- [x] Assignee
- [x] Labels
- [x] Link to epic: If belonging to an epic
- [x] A task is always a separate ticket/issue
- [x] UI/UX Design (optional)
- [x] Blockers/Prerequisites (optional): Use Zenhub blocked by/blocking feature

### LINKED ISSUE #1747: [SUBGRAPH BUG] flowDistributionUpdatedEvents not being picked up by the graph
**Describe the bug:**
A clear and concise description of what the bug is. What is the data you are having issues with?

**Subgraph endpoint:** [e.g. https://thegraph.com/hosted-service/subgraph/superfluid-finance/protocol-v1-goerli]
https://api.thegraph.com/subgraphs/name/superfluid-finance/protocol-v1-optimism-goerli

**Query:**
e.g. 
```
query MyQuery {
  flowDistributionUpdatedEvents {
    id
  }
}
```

**Expected data:**
I would expect to see more than 1 flowDistributionUpdated events, e.g.
https://goerli-optimism.etherscan.io/tx/0xa00990ffad4bd6c2482ccf3070a02c9be21828acc11e3075119e84f5991b59bc#eventlog
https://goerli-optimism.etherscan.io/tx/0xdb60cd7725c48ee8dd6f72ccfc38233bcc8631791fca0e2a9929dfdf76d808ba#eventlog
https://goerli-optimism.etherscan.io/tx/0xdb60cd7725c48ee8dd6f72ccfc38233bcc8631791fca0e2a9929dfdf76d808ba#eventlog

**Actual data:**
A clear and concise description of what is incorrect with the data or paste the returned data (within reason).
Only one distribution updated event is shown
Screenshot 2023-11-14 at 12 20 13

**Additional context:**
This wallet has got a bunch of events https://goerli-optimism.etherscan.io/address/0xdc305dd2551a8d331d372460ed72e6ad846f435a

### LINKED ISSUE #1749: [SUBGRAPH BUG] Nulls returned when querying "poolDistributor" on flowDistributionUpdatedEvents
**Describe the bug:**
Cannot query "poolDistributor" on flowDistributionUpdatedEvents , it will return an error - "Null value resolved for non-null field `poolDistributor`"

**Subgraph endpoint:**
All of the networks where GDA is deployed

**Query:**
```
query {
  flowDistributionUpdatedEvents {
    id
    poolDistributor {
      account {
        id
      }
    }
  }
}
```

**Expected data:**
Expecting to see all the information about the pool distributor

**Actual data:**
"Null value resolved for non-null field `poolDistributor`"

### PR DESCRIPTION
**Pull request by 0xdavinchee**
### Changes For Trail of Bits Audit Review
- Fix high severity vulnerability which would not allow liquidators to close insolvent streams in pools where `distributionFromAnyAddress` config is set to false: https://github.com/superfluid-finance/protocol-monorepo/pull/1729/files#diff-f22822bca4f06e61ac121310093c380eea97c2b750debd1b48f5382e3be774c6R459
- Address emitting of `PoolConnectionUpdated` when no changes were made (**TOB-SUPERFLUID-2**)
- Address safe casting of buffer to uint96 (**TOB-SUPERFLUID-5**)
- Address off-by-one gas left check (**TOB-SUPERFLUID-6**)
- Address ownership transfer of beacon ownership to governance
- Add `updatePoolBeaconLogic` to host and governance
- fix flakey tests
- uses timestamp from currentContext
- uses timestamp from` ISuperfluid(superToken.getHost()).getNow();` instead of `block.timestamp`
- Removal of functions and constants in `GeneralDistributionAgreementV1.sol` for max contract size limit restriction
- Add custom error hashes to `IPoolNFTBase.sol`

### CODE REVIEW

File: packages/ethereum-contracts/contracts/agreements/gdav1/GeneralDistributionAgreementV1.sol [hunk 1]
```diff
@@ -467,7 +463,10 @@ contract GeneralDistributionAgreementV1 is AgreementBase, TokenMonad, IGeneralDi
 
         newCtx = ctx;
 
-        if (!pool.distributionFromAnyAddress()) {
+        // we must check if the requestedFlowRate is greater than 0 here
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/gdav1/GeneralDistributionAgreementV1.sol:456]: Let's improve our echidna fuzzing with better assertions on expected liquidations.

File: packages/ethereum-contracts/contracts/agreements/gdav1/GeneralDistributionAgreementV1.sol [hunk 2]
```diff
@@ -338,31 +330,32 @@ contract GeneralDistributionAgreementV1 is AgreementBase, TokenMonad, IGeneralDi
...
                 (, PoolMemberData memory poolMemberData) = _getPoolMemberData(token, msgSender, pool);
                 token.terminateAgreement(_getPoolMemberHash(msgSender, pool), 1);
 
                 _clearPoolConnectionsBitmap(token, msgSender, poolMemberData.poolID);
             }
         }
 
-        emit PoolConnectionUpdated(token, pool, msgSender, doConnect, currentContext.userData);
+        // we only emit PoolConnectionUpdated if the connection state has changed
+        if (doConnect && !isConnected || !doConnect && isConnected) {
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/agreements/gdav1/GeneralDistributionAgreementV1.sol]: do `doConnect != isConnected`

File: packages/ethereum-contracts/contracts/agreements/gdav1/GeneralDistributionAgreementV1.sol [hunk 3]
```diff
@@ -823,7 +805,7 @@ contract GeneralDistributionAgreementV1 is AgreementBase, TokenMonad, IGeneralDi
         data = new bytes32[](2);
         data[0] = bytes32(
             (uint256(int256(uIndexData.flowRate)) << 160) | (uint256(uIndexData.settledAt) << 128)
-                | (uint256(uIndexData.totalBuffer) << 32) | (uIndexData.isPool ? 1 : 0)
+                | (uint256(uIndexData.totalBuffer.toUint96()) << 32) | (uIndexData.isPool ? 1 : 0)
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/agreements/gdav1/GeneralDistributionAgreementV1.sol:798]: cc @hellwolf please take a closer look

**Comment by 0xdavinchee** [comment]
`block.timestamp` should not be used even in view functions as we want to refer to time determined by the host

File: packages/ethereum-contracts/contracts/interfaces/agreements/gdav1/IPoolNFTBase.sol [hunk 4]
```diff
@@ -4,14 +4,14 @@ pragma solidity >=0.8.4;
...
-    error POOL_NFT_NOT_REGISTERED_POOL();
-    error POOL_NFT_TRANSFER_NOT_ALLOWED();
-    error POOL_NFT_TRANSFER_CALLER_NOT_OWNER_OR_APPROVED_FOR_ALL();
+    error POOL_NFT_APPROVE_TO_CALLER();                             // 0x9212b333
+    error POOL_NFT_ONLY_SUPER_TOKEN_FACTORY();                      // 0x1fd7e3d8
+    error POOL_NFT_INVALID_TOKEN_ID();                              // 0x09275994
+    error POOL_NFT_APPROVE_TO_CURRENT_OWNER();                      // 0x020226d3
+    error POOL_NFT_APPROVE_CALLER_NOT_OWNER_OR_APPROVED_FOR_ALL();  // 0x1e82f255
+    error POOL_NFT_NOT_REGISTERED_POOL();                           // 0x6421912e
+    error POOL_NFT_TRANSFER_NOT_ALLOWED();                          // 0x432fb160
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/interfaces/agreements/gdav1/IPoolNFTBase.sol]: extra space

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidGovernance.sol [hunk 5]
```diff
@@ -41,7 +41,8 @@ interface ISuperfluidGovernance {
         ISuperfluid host,
         address hostNewLogic,
         address[] calldata agreementClassNewLogics,
-        address superTokenFactoryNewLogic
+        address superTokenFactoryNewLogic,
+        address beaconNewLogic
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidGovernance.sol:45]: @d10r we need to upgrade the governance contract, right?

File: packages/ethereum-contracts/test/foundry/gov/SuperfluidGovernanceII.t.sol [hunk 6]
```diff
@@ -47,6 +49,39 @@ contract SuperfluidGovernanceIntegrationTest is FoundrySuperfluidTester {
...
+
+        vm.expectRevert();
+        sf.governance.updateContracts(sf.host, address(0), new address[](0), address(0), address(newPoolLogic));
+    }
+
+    function testUpdateContractsToUpgradePoolBeaconLogic() public {
+        SuperfluidPool newPoolLogic = new SuperfluidPool(sf.gda);
+        vm.startPrank(sf.governance.owner());
+        sf.governance.updateContracts(sf.host, address(0), new address[](0), address(0), address(newPoolLogic));
+        vm.stopPrank();
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/test/foundry/gov/SuperfluidGovernanceII.t.sol:63]: assertEq(newPoolLogic, gda.beacon.implementation())

**Review by hellwolf — COMMENTED** [review]
left some comments

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidGovernance.sol (hunk 5 above)
Comment (d10r) [reply packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidGovernance.sol:45]: yes

**Review by d10r — CHANGES_REQUESTED** [review]
After spending a lot of time with the GDA as an integrator, I have come to the point where I think I can concisely explain my beef with the current semantics of flow distributions related to the handling of division remainders (adjustmentFlow, but not only). Here it is: https://github.com/d10r/sf-gda-issues
I'm not sure how much of the related logic is determined by the GDA contract vs semantic-money lib @hellwolf .

**Comment by hellwolf** [comment]
Please clarify the context of the problem beyond the single word "flawed". It will need me some time follow your examples otherwise; perhaps more efficiently in a discussion session instead.

There were certainly some trade off, we have indeed talked about it many times, perhaps needing a bit more write-up for people developing on GDA closely. However if there are problem resulting impossibility of certain important use case, then that's something more important to address.

**Comment by d10r** [comment]
> Please clarify the context of the problem beyond the single word "flawed"

I extensively elaborated [here](https://github.com/d10r/sf-gda-issues/blob/master/test/GDASemanticIssues.t.sol).

**Comment by hellwolf** [comment]
ah, missed it, let's talk later this week or so

**Comment by 0xdavinchee** [comment]
merging this PR as of now, but opened a new issue [here](https://github.com/superfluid-finance/protocol-monorepo/issues/1772) for tracking reported friction as an integrator of the GDA

cc @d10r

**Pull request merged**